python/tests/test_topology.py

```diff
@@ -5414,6 +5616,143 @@ def test_many_trees_recurrent_mutations_internal_samples(self):
+    """
+    Tests for the keep_input_roots option to simplify.
+    """
+
+    def verify(self, ts):
+        # Called by the examples in ExampleTopologyMixin
+        samples = ts.samples()
+        self.verify_keep_input_roots(ts, samples[:2])
+        # self.verify_keep_input_roots(ts, samples[:3])
+        # self.verify_keep_input_roots(ts, samples[:-1])
+        # self.verify_keep_input_roots(ts, samples)
+
+    def verify_keep_input_roots(self, ts, samples):
+        ts_with_roots, node_map = self.do_simplify(
+            ts, samples, keep_input_roots=True, filter_sites=False, compare_lib=True
+        )
+        new_to_input_map = {
+            value: key for key, value in enumerate(node_map) if value != tskit.NULL
+        }
+        for (left, right), input_tree, tree_with_roots in tsutil.coiterate(
+            ts, ts_with_roots
+        ):
+            input_roots = input_tree.roots
+            self.assertGreater(len(tree_with_roots.roots), 0)
+            for root in tree_with_roots.roots:
+                # Check that the roots in the current
+                input_root = new_to_input_map[root]
+                self.assertIn(input_root, input_roots)
+                input_node = ts.node(input_root)
+                new_node = ts_with_roots.node(root)
+                self.assertEqual(new_node.time, input_node.time)
+                self.assertEqual(new_node.population, input_node.population)
+                self.assertEqual(new_node.individual, input_node.individual)
+                self.assertEqual(new_node.metadata, input_node.metadata)
+                # This should only be marked as a sample if it's an
+                # element of the samples list.
+                self.assertEqual(new_node.is_sample(), input_root in samples)
+                # Find the MRCA of the samples below this root.
+                root_samples = list(tree_with_roots.samples(root))
+                mrca = functools.reduce(tree_with_roots.mrca, root_samples)
+                if mrca != root:
+                    # If the MRCA is not equal to the root, then there should
+                    # be a unary branch joining them.
+                    self.assertEqual(tree_with_roots.parent(mrca), root)
+                    self.assertEqual(tree_with_roots.children(root), (mrca,))
+
+                    # Any mutations that were on the path from the old MRCA
+                    # to the root should be mapped to this node.
+                    u = new_to_input_map[mrca]
+                    root_path = []
+                    while u != tskit.NULL:
+                        root_path.append(u)
+                        u = input_tree.parent(u)
+                    input_sites = {site.position: site for site in input_tree.sites()}
+                    new_sites = {
+                        site.position: site for site in tree_with_roots.sites()
+                    }
+                    positions = set(input_sites.keys()) & set(new_sites.keys())
```

Comment:
shouldn't these be the same, because of `filter_sites=False`?

Reply:
Oh, never mind, I see why. But I guess that we could make sure we don't skip sites by doing
```
                    input_sites = {site.position: site for site in input_tree.sites()
                            if site.position >= left and site.position < right}
                    new_sites = {
                        site.position: site for site in tree_with_roots.sites()
                        if site.position >= left and site.position < right
                    }
                    self.assertEqual(set(input_sites.keys()), set(new_sites.keys()))
                    positions = input_sites.keys()
```

Reply:
(I stuck this in my PR to this branch.)

Reply:
Good idea.